libtiledbsoma/src/utils/fastercsx.h

```diff
@@ -0,0 +1,545 @@
+void count_rows_(
+    ThreadPool* const tp,
+    uint64_t n_row,
+    uint64_t nnz,
+    std::vector<tcb::span<COO_IDX const>> const& Ai,
+    tcb::span<CSX_MAJOR_IDX>& Bp) {
+    assert(nnz == sum_over_size_(Ai));
+    assert(Bp.size() == n_row + 1);
+    assert(tp->concurrency_level() >= 1);
+
+    std::fill(Bp.begin(), Bp.end(), 0);
+
+    auto partitions = partition_views_(
+        Ai, tp->concurrency_level(), 32 * MEBI /* heuristic (empirical) */);
+    auto n_partitions = partitions.size();
+    if (n_partitions > 1) {
+        // for multiple partitions, allocate accumulators and perform in
+        // parallel
+        std::vector<std::vector<CSX_MAJOR_IDX>> partition_counts(
+            n_partitions, std::vector<CSX_MAJOR_IDX>(n_row + 1, 0));
+
+        auto status = parallel_for(
+            tp,
+            0ul,
+            n_partitions,
+            [&partition_counts, &partitions, &n_row](const uint64_t partition) {
+                auto& counts = partition_counts[partition];
+                for (auto& Ai_view : partitions[partition].views) {
+                    for (size_t n = 0; n < Ai_view.size(); n++) {
+                        auto row = Ai_view[n];
+                        if ((row < 0) ||
+                            (static_cast<std::make_unsigned_t<COO_IDX>>(row) >=
+                             n_row)) [[unlikely]]
+                            throw std::out_of_range("Coordinate out of range.");
+                        counts[row]++;
+                    }
+                }
+                return Status::Ok();
+            });
+        assert(status.ok());
+
+        // sum partitioned counts
+        for (uint64_t partition = 0; partition < n_partitions; ++partition) {
+            for (uint64_t row = 0; row < n_row + 1; ++row) {
+                Bp[row] += partition_counts[partition][row];
+            }
+        }
+    } else if (n_partitions > 0) {
+        // for a single partition, just accumulate directly into the output
+        // array
+        assert(Ai.size() == 1);
+        assert(
+            Ai[0].data() == partitions[0].views[0].data() &&
+            Ai[0].size() == partitions[0].views[0].size());
+        auto& Ai_view = Ai[0];
+        for (uint64_t n = 0; n < nnz; ++n) {
+            uint64_t row = Ai_view[n];
+            if ((row > n_row - 1) || (row < 0)) [[unlikely]]
+                throw std::out_of_range("Coordinate out of range.");
+            Bp[row]++;
+        }
+    }
+    // else, zero length array
+}
+
+/**
+ * @brief Sum counts in place to create index pointers. Same as a prefix sum,
+ * but with the result stored offset by one position (ie., first position is
+ * always zero).
+ */
+template <typename CSX_MAJOR_IDX>
+void sum_rows_to_pointers_(tcb::span<CSX_MAJOR_IDX>& Bp) {
+    CSX_MAJOR_IDX cumsum = 0;
+    for (uint64_t i = 0; i < Bp.size(); i++) {
+        auto temp = Bp[i];
+        Bp[i] = cumsum;
+        cumsum += temp;
+    }
```

Comment:
```suggestion
    std::partial_sum(Bp.begin(), Bp.end(), Bp.begin());
```

Reply:
As I read the spec for std::partial_sum, this won't work for two reasons:
* The suggestion is not equivalent to the original code.   Given [1,2,3], the original produces [0, 1, 3, 6].  Partial sum produces [1, 3, 6].
* We also can't do `std::partial_sum(Bp.begin(), Bp.end()-1, Bp.begin()+1)`, as the output iterator overlaps the input - e.g., it would produce undefined behavior given the std::partial_sum spec..

Reply:
Yeah you're right, it is not equivalent. Will mark this as resolved.